### PR TITLE
Add solutions for lab01

### PR DESCRIPTION
**Pull request by patrick-nanys**

### CODE REVIEW

File: lab_solutions/lab01_reduce.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+# -*- coding: utf-8 -*-
+#
+# Copyright © 2020 Patrick Nanys <patrick.nanys2000@gmail.com>
+#
+# Distributed under terms of the MIT license.
+
+
+def reduce(l, acc_func, accumulator=None):
+    if accumulator is not None:
+        l.insert(0, accumulator)
```
Comment (juditacs) [inline lab_solutions/lab01_reduce.py]: You should not modify `l`. It's passed as a reference so the original list is the one that's modified here.

File: lab_solutions/lab01_reduce.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    reduce(l1, add)
+
+    assert (reduce(l1, add) == 7)
+    assert (reduce(l1, add, 10) == 17)
+
+    l2 = ["foo", "bar", "hello"]
+
+    assert (reduce(l2, string_len_add, 0) == 11)
+
```
Comment (juditacs) [inline lab_solutions/lab01_reduce.py:47]: Please add a test for finding the maximum of a list.

Add another test that checks if the original list is unchanged (this should fail with your current solution).

File: lab_solutions/lab01_qsort.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+# Distributed under terms of the MIT license.
+
+
+def swap(l, idx1, idx2):
+    temp = l[idx1]
+    l[idx1] = l[idx2]
+    l[idx2] = temp
+
+
+def qsort(l, start, end):
```
Comment (juditacs) [inline lab_solutions/lab01_qsort.py]: `qsort` should only take one mandatory argument, the list.

**Review by juditacs — CHANGES_REQUESTED** [review]
Nice work.

I found a few smaller issues, please address them.

File: lab_solutions/lab01_qsort.py (hunk 3 above)
Comment (patrick-nanys) [reply lab_solutions/lab01_qsort.py]: Sorry, I accidentally added a previous version of my code.

File: lab_solutions/lab01_qsort.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    for i in range(start + 1, end):
+        current_idx = i
+        if l[current_idx] < l[pivot_idx]:
+            while current_idx > pivot_idx:
+                swap(l, current_idx - 1, current_idx)
+                current_idx -= 1
+            pivot_idx += 1
+
+    # call next partitions
+
```
Comment (juditacs) [inline lab_solutions/lab01_qsort.py]: no need for so many newlines. It seems like a small thing but it makes the function harder to fit in one screen.

**Review by juditacs — CHANGES_REQUESTED** [review]
One small remark, please fix it and then I'll approve this PR.

**Comment by juditacs** [comment]
Thanks. Good work.
